File: Data Structures & Algorithms/house-robber-ii/submission-0.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n= nums.size();
        if(n==1)
            return nums[0];

        vector<vector<int>> dp(2,vector<int> (n,0));
        int res;

        res = max(nums[0],nums[1]);

        dp[0][1] = nums[1];
        int x,y;
        
        for(int i=2;i<nums.size();i++){
           dp[0][i] = max(dp[0][i-1], dp[0][i-2] + nums[i]);
           res = max(res,dp[0][i]);
        }

        dp[1][0] = nums[0];
        dp[1][1] = max(nums[0],nums[1]);

        for(int i=2;i<n-1;i++){
            dp[1][i] = max(dp[1][i-1], dp[1][i-2] + nums[i]);
            res = max(res,dp[1][i]);
        }


        return res;
    }
};

File: Data Structures & Algorithms/sliding-window-maximum/submission-0.java
class Solution {


    public int[] maxSlidingWindow(int[] nums, int k) {
        
        Map<Integer,Integer> mp = new TreeMap<>();
        int sz=0;
        int key,value;
        List<Integer> res= new ArrayList<>();
        for(int i=0;i<nums.length;i++){
            sz++;
            key = -nums[i];
            mp.put(key, mp.getOrDefault(key,0)+1);

            if(sz > k){
                key = -nums[i-k];
                value = mp.get(key);

                if((value - 1)==0)
                    mp.remove(key);
                else mp.put(key,value-1);  

                res.add(-mp.keySet().iterator().next());
            }else if(sz == k){
                res.add(-mp.keySet().iterator().next());
            }

            //
        }

        int[] arr = res.stream().mapToInt(Integer::intValue).toArray();

        return arr;
    }
}
